doc/SAI-Proposal-ARS.md

```diff
@@ -0,0 +1,1327 @@
+3. Dynamic path assignment and reassignment without any interaction with the control plane
+4. Minimize traffic drop when link flaps by HW based link down detection and path reassignment
+
+### 3.1 Terms and Acronyms
+| Acronym   | Term      |
+|:----------|:----------|
+| Micro Flow    | A single traffic flow typically identified by 5 tuple    |
+| Macro Flow    | collection of micro flows that hash to same value   |
+| Idle Time | Time when there are no packets observed within a macro flow |
+| Flowlet    | contagious subsequence of packets within a macro flow separated by an idle time interval    |
+| Macro Flow Table    | HW table to store per macro flow state    |
+| EWMA | Exponential weighted moving average |
+| ARS    | Adaptive routing and switching: Path selection is influenced by quality metrics    |
+| NHG | Next Hop Group |
+
+### 3.2 Key Idea behind ARS
+1. ARS is flowlet based
+
+      - *A flow is viewed as a sequence of flowlets separated by idle time*
+
+2. Path selection is based on local visibility into link quality
+
+      - *Two metrics are considered. Historical load:Bytes sent over an interval of time and future load: Bytes queued for transmission*
+
+3. Path reassignment takes flow (in)activity into account
+
+      - *Best path is selected when the flow starts. Path is locked while flow is active.*
+      - *Flow becomes eligible for path reassignment after the idle time.*
+
+      - *Idle time should be configured typically as 1/2 of RTT*
+
+4. Algorithm for load computation
+
+      - *EWMA is used for computing the historical and future load*
+
+5. Quantization
+
+      - *Computed value need to be quantized into buckets or bands for easier intepretation*
+
+6. Non minimal cost path selection
+
+      - *ARS can be used to partition a NHG into a primary and alternate set of paths. Alternate paths are typically non minimal cost or side links based on the topology. The alternate paths are chosen when primary paths are relatively worse.*
+
+7. LAG
+
+      - *Besides being able to choose best quality path from the NHG, packet should also be able to select best link available amongst LAG memembers*
+
+8. Link Failure or Admin Down
+
+      - *Besides being able to choose best quality path from the NHG, packet should also be able to select best link available amongst LAG memembers*
+
+> *Additional algorithms and quality metrics can be added laters*
+
+### 3.3 ARS path selections modes
+
+1. Flow-let based
+
+      - 	*best quality path till the flow-let hits an idle time*
+
+      - *random path with out considering path quality*
+
+2. Per packet
+
+      - *best quality path assignment for each packet*
+
+      - *random path assignment for each packet*
+
+3. Fixed: fixed path assignment
+
+      - *mostly used for debugging*
+
+## 4.0 ARS Packet Flow
+
+![Adaptive Routing and Switching](../figures/ARS-Packet-Flow.png "Figure 1: ARS Packet Flow")
+

```

Comment:
This diagram probably needs a bit of a cleanup and some explanation. Are we storing micro-flows or macro-flows in this table? Also, is there a single logical instance of this table? Or there are n logical tables, one for each macro-flow. Also, it would help  if the color scheme used to represent a micro-flows is consistent throughout the diagram.

Reply:
Thanks for this. Cleaned it up in the new PR.